client/db/src/lib.rs

```diff
@@ -1307,11 +1332,32 @@ impl<Block: BlockT> Backend<Block> {
 }
 
 fn apply_state_commit(transaction: &mut DBTransaction, commit: sc_state_db::CommitSet<Vec<u8>>) {
-	for (key, val) in commit.data.inserted.into_iter() {
-		transaction.put(columns::STATE, &key[..], &val);
-	}
-	for key in commit.data.deleted.into_iter() {
-		transaction.delete(columns::STATE, &key[..]);
+	let mut key_buffer = Vec::new();
+	for child_data in commit.data.into_iter() {
+		if child_data.0.is_top_trie() {
+			// empty prefix
+			for (key, val) in child_data.1.inserted.into_iter() {
+				transaction.put(columns::STATE, &key[..], &val);
+			}
+			for key in child_data.1.deleted.into_iter() {
+				transaction.delete(columns::STATE, &key[..]);
+			}
+		} else {
+			let keyspace = child_data.0.keyspace();
+			let keyspace_len = keyspace.len();
+			key_buffer.resize(keyspace_len, 0);
```

Comment:
There should exists a function for this. Better would be a type `ChildKey` or whatever.

```
impl ChildKey {
    fn new(keyspace, key)
    fn update_key()
    fn update_space_and_key()
}
```

Something in this direction.